sklearn/tree/_criterion.pyx

```diff
@@ -79,22 +79,22 @@ cdef class Criterion:
 
         pass
 
-    cdef void reset(self) nogil:
+    cdef int reset(self) nogil except -1:
```

Comment:
I think it's fine have those C functions to use an int return code that is used to flag memory allocation error. It can indeed help Cython reduce it's overhead.